---
- 07/23

In [128]:
import pandas as pd
import numpy as np

#워닝 메시지
import warnings
warnings.filterwarnings('ignore')

In [129]:
df_외감=pd.read_csv('./Dataset/외감raw3.csv',encoding='euc-kr')
df_외감.drop(['회사명.1','회계년도.1'],axis=1,inplace=True)

In [130]:
df_외감.shape

(295975, 54)

In [131]:
len(df_외감['회사명'].unique())

51667

- shape : (295975,54) /  기업수 : 51667개
---

In [132]:
list=['SK리츠운용주식회사', '굿모닝대신리츠주식회사', '(주)동방리츠', '대영리츠건설(주)', '드림리츠(주)',
       '(주)리츠라인', '(주)리츠산업', '(주)리츠홀딩스', '(주)블리츠웨이', '메리츠건설(주)',
       '(주)서울리츠임대주택제2호위탁관리부동산투자회사', '신한리츠운용주식회사',
       '엔에이치농협리츠운용주식회사', '영신리츠팜주식회사', '(주)제이알리츠', '월드리츠개발주식회사',
       '주식회사경기리츠공공임대제1호위탁관리부동산투자회사', '주식회사경성리츠', '주식회사글리츠리츠',
       '주식회사대신케이리츠물류1호위탁관리부동산투자회사', '주식회사리츠파크', '주식회사무등리츠', '주식회사비앤에스리츠',
       '주식회사서울리츠임대주택제1호위탁관리부동산투자회사', '주식회사서울리츠임대주택제3호위탁관리부동산투자회사',
       '주식회사서울리츠임대주택제4호위탁관리부동산투자회사', '주식회사아이에이치큐리츠', '주식회사와이에스리츠',
       '주식회사제로에너지임대리츠2호위탁관리부동산투자회사', '주식회사조일리츠', 
       '주식회사토지지원리츠제1호위탁관리부동산투자회사', '주식회사토지지원리츠제2호위탁관리부동산투자회사', '주식회사평안리츠',
       '주식회사플리츠', '주식회사하우리츠', '주식회사힘내라리츠', '(주)청리츠앤파트너스',
       '(주)케이리츠투자운용', '케이리츠주식회사', '켄달스퀘어리츠운용주식회사', '한국리츠주식회사']
for i in list:
       df_외감 = df_외감[~df_외감['회사명'].str.contains(i)]


In [133]:
list =['HSBC펀드서비스(주)', '주식회사어니스트펀드','주식회사인천투자펀드2', '주식회사피플펀드컴퍼니']

for i in list:
    df_외감 = df_외감[~df_외감['회사명'].str.contains(i)]
    

In [134]:
list=['(주)건설투자', '(주)국제선박투자운용','(유)비티투자','스마트투자파트너스주식회사', '(주)에스와이투자자문', '신동해투자일호피에프브이(주)',
       '아이비투자검단(주)', '아이비투자을지로주식회사', '에이스투자(주)', '에이치티투자금융주식회사',
       '엠제이투자시스템주식회사', '유한회사한일자산관리앤투자', '인터씨티개발투자(주)', '일조투자디앤씨(주)',
       '주식회사건인투자', '주식회사드림부동산투자연구소', '주식회사세원투자건업', '주식회사세원투자산업',
       '주식회사아프로투자', '주식회사에이디투자', '주식회사에이치엠지기술투자',
       '주식회사이상투자그룹', '주식회사제이에스투자원', '주식회사케이티투자운용',
       '주식회사테라투자파트너스', '주식회사한국민간인프라투자', '주식회사한우즈벡실린더투자', '주식회사화신투자',
       '(주)투자가이드', '케이디기술투자(주)', '케이엔씨개발투자컨설팅', '코레이트투자운용주식회사', '(주)한진정석투자',
       '(주)현대부동산지분투자', '포항프라이머스프로젝트투자금융주식회사', '한원광장프로젝트투자금융주식회사']
for i in list:
    df_외감 = df_외감[~df_외감['회사명'].str.contains(i)]
    

In [135]:
df_외감 = df_외감[~df_외감['회사명'].str.contains('부동산투자회사')]
df_외감 = df_외감[~df_외감['회사명'].str.contains('금융투자주식회사')]
df_외감 = df_외감[~df_외감['회사명'].str.contains('선박투자회사')]
df_외감 = df_외감[~df_외감['회사명'].str.contains('투자회사')]
df_외감 = df_외감[~df_외감['회사명'].str.contains('투자주식회사')]
df_외감 = df_외감[~df_외감['회사명'].str.contains('투자개발')]
df_외감 = df_외감[~df_외감['회사명'].str.contains('부동산투지회사')]
df_외감 = df_외감[~df_외감['회사명'].str.contains('금융투자')]

In [136]:
df_외감.isna().sum()

회사명                                       0
거래소코드                                     0
회계년도                                      0
통계청 한국표준산업분류 코드 10차(대분류)                  0
부채비율                                     80
당좌비율                                     80
유동비율                                     80
이자보상배율(이자비용)                             80
영업활동으로 인한 현금흐름(간접법)(*)(IFRS연결)(천원)       69
이자비용(IFRS연결)(천원)                         69
차입금의존도                                   80
자기자본구성비율                                 80
당기순이익(손실)(IFRS연결)(천원)                    69
감가상각비(백만원)                             2000
금융비용(백만원)                               300
단기차입금(*)(IFRS연결)(천원)                     69
유동성장기부채(*)(IFRS연결)(천원)                   69
장기차입금(*)(IFRS연결)(천원)                     69
법인세비용차감전(계속사업)손익(백만원)                   103
매출액영업이익률                                 87
자기자본순이익률                                 87
총자본순이익률                                  87
매출액(수익)(*)(IFRS연결)(천원)          

---
# traget 레이블링   
    - 감사의견 코드
    - 자본잠식률
    - 회생절차 기업
    
1. 감사의견코드

In [137]:
df_외감['감사의견코드']=df_외감['감사의견코드'].map({
    'AG': 1,
    'DS':1,
    'DI':1,
    'DU':1,
    'QA':0,
    'QC':0,
    'QG':0,
    'QQ':0,
    'QS':1,
    'QU':0,
    'UQ':0,
    'ZZ':1
})

 2. 자본잠식  
    - 컬럼을 하나 만듬
    - 조건에 따라 레이블링
    

In [138]:
df_외감['자본잠식률']=(df_외감['자본금(*)(IFRS연결)(천원)'] - df_외감['자본(*)(IFRS연결)(천원)'])/df_외감['자본금(*)(IFRS연결)(천원)']

In [139]:
mask = (df_외감['거래소코드'] == df_외감['거래소코드'].shift(-1)) & (df_외감['자본잠식률'] >= 50)

In [140]:
df_외감['감사의견코드'].value_counts()

감사의견코드
0.0    239093
1.0     20893
Name: count, dtype: int64

In [141]:
df_외감.loc[mask, '감사의견코드'] = 1.0

In [142]:
mask.value_counts()

False    289149
True       4809
Name: count, dtype: int64

In [143]:
df_외감['감사의견코드'].value_counts()

감사의견코드
0.0    236292
1.0     24748
Name: count, dtype: int64

In [144]:
24748 - 20893

3855

- 자본잠식 > 50인 3855개가 0에서 1로 레이블링 변경

3. 회생절차 기업 확인하기!

In [145]:
# 회생절차 개시신청
df_외감.loc[132330,'감사의견코드']=1.0
df_외감.loc[12111,'감사의견코드']=1.0
df_외감.loc[16614,'감사의견코드']=1.0
df_외감.loc[77037,'감사의견코드']=1.0
df_외감.loc[191585,'감사의견코드']=1.0
df_외감.loc[77037,'감사의견코드']=1.0
df_외감.loc[211114,'감사의견코드']=1.0
df_외감.loc[228410,'감사의견코드']=1.0
df_외감.loc[256817,'감사의견코드']=1.0
df_외감.loc[280187,'감사의견코드']=1.0
df_외감.loc[288076,'감사의견코드']=1.0
# 해산사유 발생
# 77485 (주)세실 2017년 파산  ->2016년 레이블링 1.0으로 수정 필요


In [146]:
df_외감['감사의견코드'].value_counts()

감사의견코드
0.0    236282
1.0     24758
Name: count, dtype: int64

In [147]:
df_외감['감사의견코드'].isna().sum()

32918

- 0.0 : 236282
- 1.0 : 24758
- 결측치 : 32918

---

> 오름차순으로 정렬 회사명, 거래소 코드로

In [148]:
df_외감.sort_values(by=['회사명','거래소코드'], ascending=True,inplace=True)

> 회사명이 한개인 기업들 삭제
    - 2369개 삭제

In [149]:
len(df_외감[df_외감['거래소코드'].map(df_외감['거래소코드'].value_counts()) == 1])


2369

In [150]:
df_외감=df_외감[df_외감['거래소코드'].map(df_외감['거래소코드'].value_counts()) != 1]


In [151]:
df_외감['감사의견코드'].value_counts()

감사의견코드
0.0    234559
1.0     24140
Name: count, dtype: int64

> 파생변수 : Year 만들기

In [152]:
df_외감['Year']= df_외감['회계년도'].replace('/','').astype(str).str[:4].astype(int)

In [153]:
# pd.reset_option('display.max_rows', None)
# df_외감[(df_외감['거래소코드'] == df_외감['거래소코드'].shift(-1)) & (df_외감['Year'].diff(-1) != -1) &(df_외감['감사의견코드'] ==1) &(df_외감['Year'] !=2012)]

> 회사 이름이 같고, 연도 차이가 일년만 나는 데이터프레임 확인!

> 만약 2017년에 부실징후 포착 -> 2016년에 레이블링을 해야하는데 , 2017전년도 데이터가 2014년 같이 불연속인경우 고려

In [154]:
df_외감[~(df_외감['회사명'] == df_외감['회사명'].shift(-1)) | (abs(df_외감['Year']-df_외감['Year'].shift(-1)) == 1)]['감사의견코드'].value_counts()

감사의견코드
0.0    231716
1.0     23583
Name: count, dtype: int64

In [155]:
# df_외감.to_csv('t시점_외감.csv',index=False,encoding='euc-kr')

----


- df_외감 : (291589,55)
- 감사의견코드 결측치 : 32890
- 감사의견코드 value_counts
    - true  : 234559
    - false : 24140

---

* 기업의 target : 0 이면 --> 그 기업 데이터 모두 유지
* 기업의 target : 1 이면 --> 그 기업 1 이전 데이터 유지 & 1 이후 데이터 삭제

In [156]:
import pandas as pd

# 거래소코드로 그룹화하여 처리하는 함수
def process_group(group):
    # 빈 그룹인 경우 그대로 반환
    if group.empty:
        return group
    
    # 거래소코드 그룹별로 처리
    if 1 in group['감사의견코드'].values:
        # 1이 처음으로 나오는 인덱스 찾기
        first_index = group.index[group['감사의견코드'] == 1][0]
        
        # 처음으로 1이 나오는 행의 아래 행들 삭제
        group = group.loc[:first_index]
    else:
        # 1이 없는 경우 그룹 그대로 유지
        pass
    
    return group



In [157]:
# '거래소코드'로 그룹화하여 함수를 적용하고 다시 합치기
processed_df = df_외감.groupby('거래소코드').apply(process_group)

In [158]:
processed_df.shape

(247790, 56)

In [159]:
processed_df['감사의견코드'].value_counts()

감사의견코드
0.0    199447
1.0     16452
Name: count, dtype: int64

In [160]:
processed_df['감사의견코드'].isna().sum()

31891

- 기업의 처음 1이 나온 이후의 데이터 삭제!
    - shape(247790,56)
    - 0.0 : 199447
    - 1.0 : 16452
    - 결측치 : 31891

* 기업의 개수가 1개인 기업은 drop!

In [176]:
processed_df=processed_df[processed_df['거래소코드'].map(processed_df['거래소코드'].value_counts()) != 1]


In [179]:
processed_df.shape

(244823, 56)

In [180]:
processed_df['감사의견코드'].value_counts()

감사의견코드
0.0    199447
1.0     13485
Name: count, dtype: int64

In [182]:
processed_df['감사의견코드'].isna().sum()

31891

- 0.0 : 199447
- 1.0 : 13485
- 결측치 : 30891
- 총 2967개 감소

---

#### 거래소 코드가 같은 회사라면, t-1감사의견코드라는 행에 다음년도 감사의견코드 대입

In [183]:
mask = processed_df['거래소코드'] == processed_df['거래소코드'].shift(-1)
processed_df['t-1감사의견코드'] = np.where(mask,processed_df['감사의견코드'].shift(-1), np.nan)

In [184]:
processed_df['t-1감사의견코드'].value_counts()

t-1감사의견코드
0.0    182115
1.0     13485
Name: count, dtype: int64

In [185]:
processed_df.dropna(subset='t-1감사의견코드',axis=0,inplace=True)

In [186]:
processed_df.drop(columns='감사의견코드',axis=1,inplace=True)

In [187]:
processed_df.isna().sum()

회사명                                      0
거래소코드                                    0
회계년도                                     0
통계청 한국표준산업분류 코드 10차(대분류)                 0
부채비율                                    60
당좌비율                                    60
유동비율                                    60
이자보상배율(이자비용)                            60
영업활동으로 인한 현금흐름(간접법)(*)(IFRS연결)(천원)      58
이자비용(IFRS연결)(천원)                        58
차입금의존도                                  60
자기자본구성비율                                60
당기순이익(손실)(IFRS연결)(천원)                   58
감가상각비(백만원)                            1385
금융비용(백만원)                              204
단기차입금(*)(IFRS연결)(천원)                    58
유동성장기부채(*)(IFRS연결)(천원)                  58
장기차입금(*)(IFRS연결)(천원)                    58
법인세비용차감전(계속사업)손익(백만원)                   75
매출액영업이익률                                60
자기자본순이익률                                60
총자본순이익률                                 60
매출액(수익)(*)(IFRS연결)(천원)                  58
총자본회전률     

In [189]:
processed_df.drop('연구개발비(IFRS연결)(천원)',axis=1,inplace=True)

In [190]:
processed_df.columns=['회사명','거래소코드', '회계년도', '산업분류코드', '부채비율', '당좌비율', '유동비율',
       '이자보상배율', 'OCF(천원)','이자비용(천원)', '차입금의존도', '자기자본구성비율', '당기순이익(천원)',
       '감가상각비(백만원)', '금융비용(백만원)', '단기차입금(천원)','유동성장기부채(천원)', '장기차입금(천원)',
       '법인세비용차감전손익(백만원)', '매출액영업이익률', '자기자본순이익률', '총자본순이익률',
       '매출액(천원)', '총자본회전률', '자기자본회전률', '운전자본회전률', '순운전자본회전률',
       '재고자산회전률', '당좌자산회전률', '유동자산회전률', '매출액증가율', '총자본증가율', '자기자본증가율',
       '순이익증가율', '유형자산증가율', '유동자산증가율', '재고자산증가율', '영업이익증가율', '총자본투자효율',
       '부가가치율', '노동소득분배율', '자본분배율', '이윤분배율', 'ICF(천원)','FCF(천원)', '자산총계(천원)',
       '부채총계(천원)', '자본총계(천원)', 
       '산업분류', '자본금(천원)','영업손익(천원)','연구개발비(천원)','자본잠식률', 'Year', 't-1감사의견코드']

In [451]:
# processed_df.to_csv('t-1_raw_외감.csv',index=False,encoding='euc-kr')

---
- t-2 시점

In [191]:
processed_df2=processed_df[~(((processed_df['거래소코드']) ==1) & (processed_df['거래소코드'].value_counts()) == 1)]

mask = processed_df2['거래소코드'] == processed_df2['거래소코드'].shift(-1)
processed_df2['t-2감사의견코드'] = np.where(mask,processed_df2['t-1감사의견코드'].shift(-1), np.nan)

In [192]:
processed_df2['t-2감사의견코드'].value_counts()

t-2감사의견코드
0.0    146139
1.0      2298
Name: count, dtype: int64

In [193]:
processed_df2.dropna(subset='t-2감사의견코드',axis=0,inplace=True)

In [455]:
# processed_df.to_csv('t-2_raw_외감.csv',index=False,encoding='euc-kr')